Append newline to remote end's stderr when replying with an error

a-palchikov opened the pull request:
Suffix the error with a newline for added formatting when replying to a remote's side with an error.
Strip the newline when outputting the error message in tsh.

Fixes https://github.com/gravitational/teleport/issues/4542 and https://github.com/gravitational/teleport/issues/3205.

Before the change:
```
$ ssh -F ./ssh_config root@teleport-node
node is offline, please try again latersubsystem request failed on channel 0
kex_exchange_identification: Connection closed by remote host
```

After the change:
```
$ ssh -F ./ssh_config root@teleport-node
node is offline, please try again later
subsystem request failed on channel 0
kex_exchange_identification: Connection closed by remote host
```
```
$ tsh ssh root@teleport-node
error: failed connecting to node teleport-node. node is offline, please try again later
```

a-palchikov committed:
Suffix the error with a newline for added formatting when replying to a
remote's side with an error.
Strip the newline when outputting the error message in tsh.

Fixes https://github.com/gravitational/teleport/issues/4542.

a-palchikov commented on Makefile:
```diff
@@ -15,6 +15,7 @@ VERSION=5.0.0-dev
 
 DOCKER_IMAGE ?= quay.io/gravitational/teleport
 DOCKER_IMAGE_CI ?= quay.io/gravitational/teleport-ci
+DOCKER_CACHE_OPTS ?= --no-cache
```
This helps iterate faster during development by avoiding excessive image rebuilds.

a-palchikov commented on lib/srv/regular/sshserver.go:
```diff
@@ -1592,7 +1597,7 @@ func (s *Server) parseSubsystemRequest(req *ssh.Request, ctx *srv.ServerContext)
 
 func writeStderr(ch ssh.Channel, msg string) {
-	if _, err := fmt.Fprint(ch.Stderr(), msg); err != nil {
+	if _, err := io.WriteString(ch.Stderr(), msg); err != nil {
```
Changed this to use a simpler API since there is no formatting taking place.

a-palchikov commented on lib/utils/cli.go:
```diff
@@ -154,12 +154,14 @@ func UserMessageFromError(err error) string {
-			if er.Message != "" {
-				return fmt.Sprintf("error: %v", EscapeControl(er.Message))
-			}
+		if err, ok := err.(*trace.TraceErr); ok && err.Message != "" {
```
No need to use another name, the `err` will shadow `err` in outer scope and not overwrite it.

a-palchikov commented on tool/tsh/tsh.go:
```diff
@@ -908,32 +908,33 @@ func onSSH(cf *CLIConf) {
 	err = client.RetryWithRelogin(cf.Context, tc, func() error {
 		return tc.SSH(cf.Context, cf.RemoteCommand, cf.LocalExec)
 	})
-	if err != nil {
```
Is it ok to make smaller refactorings as I stumble on things?
Used an earlier return to avoid unnecessary indentation.

webvictim commented on lib/utils/cli.go:
```diff
@@ -154,12 +154,14 @@ func UserMessageFromError(err error) string {
+			return fmt.Sprintf("error: %v", EscapeControl(errMsg))
 		}
-		return fmt.Sprintf("error: %v", EscapeControl(err.Error()))
+		// Avoid escaping an error message that terminates with a newline
```
Interestingly, https://github.com/gravitational/teleport/issues/3205 may be suffering from a similar sort of formatting issue.

a-palchikov replied:
Indeed, thinking about this, it does not look like the right approach though (in the context of #3205) - it would be better to offload this to a formatting implementation that has an alternative implementation in certain cases so this happens automatically and `tsh` does not need to take any special cases into account.
I'll give more thought.

a-palchikov replied:
I think I'll go with my original changes and also propagate changes into the forwarding SSH server so it [behaves](https://github.com/gravitational/teleport/pull/4651/commits/0b24af342c484ce0dca173862924ad44313f6588) the same. The last commit also splits the `UserMessageFromError` in two implementation and uses the alternative formatting API to fix https://github.com/gravitational/teleport/issues/3205.

a-palchikov committed:
lib/utils: factor context-insensitive error formatting from UserMessageFromError.

Split UserMessageFromError in two - one with the same name that keeps
the old behavior of formatting the error via debug report if debug flag
has been specified and the other (upon it builds) to actually do error
formatting. Then use the second implementation to forward error messages
through remote channels to avoid being dependent on whether the proxy is
running with the debug flag.

r0mant commented on lib/srv/forward/sshserver.go:
```diff
@@ -1153,10 +1153,15 @@ func (s *Server) handleEnv(ch ssh.Channel, req *ssh.Request, ctx *srv.ServerCont
+	// stderr so the output does not mix with the rest of the output if the remote
+	// side is not doing additional formatting for extended data.
+	// See github.com/gravitational/teleport/issues/4542
+	payload := message + "\n"
```
Would it make sense to append `\n` only if `message` doesn't already end with it? Or it never ends with a newline?

a-palchikov replied:
I don't think this is customary to have error messages ending in a new line in teleport code base itself. This is meant for external SSH clients (actually a single client that has this behavior - OpenSSH).

awly replied:
IMO we should still add the newline conditionally, to avoid any assumptions about out error messages and blank lines on openssh side

r0mant commented on lib/utils/cli.go:
```diff
@@ -114,9 +114,44 @@ func GetIterations() int {
+
+// FormatError returns user friendly error message from error.
+// The error message is escaped if necessary
+func FormatError(err error) string {
```
Seems like the only difference b/w this new function and `UserMessageFromError` above is that the one above also returns a stack trace in debug mode? Any reason that above function can't just be used everywhere (with your new `\n` logic) instead of having two very similar ones, or am I missing anything?

a-palchikov replied:
> Seems like the only difference b/w this new function and `UserMessageFromError` above is that the one above also returns a stack trace in debug mode? Any reason that above function can't just be used everywhere (with your new `\n` logic) instead of having two very similar ones, or am I missing anything?

Yes, with `FormatError` I tried to use the same functionality in the old `UserMessageFromError` while deviating from the implicit debug prints which are problematic as shown in [this](https://github.com/gravitational/teleport/issues/3205) issue. The latter is still used universally to output errors so I did not want to alter its functionality at risk of breaking anything.

webvictim reviewed (approved):
LGTM

a-palchikov committed:
Address review comments

russjones commented on lib/utils/cli.go:
```diff
@@ -140,9 +140,44 @@ func GetIterations() int {
+	if certErr := formatCertError(err); certErr != "" {
+		return certErr
+	}
+	if log.GetLevel() == log.DebugLevel {
```
In debug mode, shouldn't we return the human readable error message as well as the stack trace?

a-palchikov replied:
Sure - it could but this was an attempt to also mitigate [this](https://github.com/gravitational/teleport/issues/3205) since the error message incl. debug stack trace popping up in the intermediary (i.e. proxy) is not going to look readable.

russjones commented on lib/srv/forward/sshserver.go:
```diff
@@ -1153,7 +1154,14 @@ func (s *Server) handleEnv(ch ssh.Channel, req *ssh.Request, ctx *srv.ServerCont
 	s.log.Error(err)
-	message := utils.UserMessageFromError(err)
+	message := utils.FormatError(err)
+	// Terminate the error with a newline when writing to remote channel's
+	// stderr so the output does not mix with the rest of the output if the remote
+	// side is not doing additional formatting for extended data.
+	// See github.com/gravitational/teleport/issues/4542
+	if !strings.HasSuffix(message, "\n") {
+		message = message + "\n"
+	}
```
Can you move this to a common function both servers can call?

russjones reviewed (commented):
Sorry about the late reply @a-palchikov , following up with old PRs.
